src/thread.rs

```diff
@@ -0,0 +1,483 @@
+//! Implements threads.
+
+use std::cell::RefCell;
+use std::convert::TryFrom;
+use std::num::NonZeroU32;
+
+use log::trace;
+
+use rustc_data_structures::fx::FxHashMap;
+use rustc_hir::def_id::DefId;
+use rustc_index::vec::{Idx, IndexVec};
+use rustc_middle::{
+    middle::codegen_fn_attrs::CodegenFnAttrFlags,
+    mir,
+    ty::{self, Instance},
+};
+
+use crate::*;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum SchedulingAction {
+    /// Execute step on the active thread.
+    ExecuteStep,
+    /// Execute destructors of the active thread.
+    ExecuteDtors,
+    /// Stop the program.
+    Stop,
+}
+
+/// A thread identifier.
+#[derive(Clone, Copy, Debug, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct ThreadId(usize);
+
+impl Idx for ThreadId {
+    fn new(idx: usize) -> Self {
+        ThreadId(idx)
+    }
+    fn index(self) -> usize {
+        self.0
+    }
+}
+
+impl From<u64> for ThreadId {
+    fn from(id: u64) -> Self {
+        Self(id as usize)
+    }
+}
+
+impl From<u32> for ThreadId {
+    fn from(id: u32) -> Self {
+        Self(id as usize)
+    }
+}
+
+impl ThreadId {
+    pub fn to_u32_scalar<'tcx>(&self) -> Scalar<Tag> {
+        Scalar::from_u32(u32::try_from(self.0).unwrap())
+    }
+}
+
+/// An identifier of a set of blocked threads.
+#[derive(Clone, Copy, Debug, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct BlockSetId(NonZeroU32);
+
+impl From<u32> for BlockSetId {
+    fn from(id: u32) -> Self {
+        Self(NonZeroU32::new(id).expect("0 is not a valid blockset id"))
+    }
+}
+
+impl BlockSetId {
+    pub fn to_u32_scalar<'tcx>(&self) -> Scalar<Tag> {
+        Scalar::from_u32(self.0.get())
+    }
+}
+
+/// The state of a thread.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum ThreadState {
+    /// The thread is enabled and can be executed.
+    Enabled,
+    /// The thread tried to join the specified thread and is blocked until that
+    /// thread terminates.
+    BlockedOnJoin(ThreadId),
```

Comment:
Interesting that this is separate from the blockset architecture. Is that a common approach? Would it increase or decrease code reuse and clarity to merge these two kinds of blocking into one?

Reply:
> Is that a common approach?

No idea. I tried looking into the actual pthread implementations, but they seem to me too optimized and error-prone for our purposes.

> Would it increase or decrease code reuse and clarity to merge these two kinds of blocking into one?

I started with `BlockSet` API because it felt like something that could be generic enough to support both Linux and Windows. However, it seems that while it allows supporting cross platform locks, it is not flexible enough to allow checking all the interesting properties of all synchronization primitives. For example, for read write locks we currently cannot check that the lock is unlocked by the same thread that locked it and for conditional variables we need a way to make other thread to acquire a lock. My plan is to have minimal support for synchronization primitives in this PR and then have a second PR that implements proper support for most synchronization primitives because this one already feels too large and hard to manage.

Reply:
So that proper support will require more scheduler changes, expanding the list here?

Please leave FIXMEs for checks that we should do but do not do yet, like the rwlock check.

> My plan is to have minimal support for synchronization primitives in this PR and then have a second PR that implements proper support for most synchronization primitives because this one already feels too large and hard to manage.

Much appreciated. :) In fact I wondered if this PR could have been shrunk by leaving out the lock changes, but maybe the basic `thread::spawn` tests already require those?

Reply:
> So that proper support will require more scheduler changes, expanding the list here?

Probably. Implementing support for all synchronization primitives is harder than I thought, so I will be able to answer that only once I figured it out.

> Please leave FIXMEs for checks that we should do but do not do yet, like the rwlock check.

The `pthread_rwlock_unlock` shim has a comment.

> In fact I wondered if this PR could have been shrunk by leaving out the lock changes, but maybe the basic thread::spawn tests already require those?

The main reason why I included locks in this PR was that we already have shims, which would be unsound if I did not fix them.